Fix consumer count off by one when closing a browser tab with a livestream that is transcoding

timminator opened the pull request:
This PR solves an issue where the consumer count is off by one when closing a browser tab with a livestream that is transcoding as described in Issue #13219. The task "CloseLivestream" is called twice in this case.

I resolved this issue by modifying the if-statement in the "CloseIfNeededAsync" task that determines if the "CloseLivestream" task is run or not. In my updated version it is not run if the livestream is transcoded, which resolves the aforementioned issue.
The consumer ount is then reduced by one when PlaybackStopped is reported, which is the expected behavior that is also seen by quitting the stream via the back button instead of closing the browser tab.

Issues:
Fixes #13219 
Fixes #13721 
Fixes #13684

gnattu commented:
This approach appears too fragile and unreliable to me. Shouldn’t those handlers verify if the stream is already closed for the consumer before proceeding, rather than assuming that the handler shouldn’t be called for any transcoded streams?

timminator commented:
Yes I agree with you completely. I also noticed that this fixes the issue but not really the underlying problem. But with my limited understanding of the code base I was not able to find a way to detect if the consumer count was already reduced by another task.
Because the issue only occurred when transcoding this was my solution.
If you have another idea or something you can point me to I'm interested in it.
My wish is just to fix this issue because it was bugging me the last few weeks until yesterday where I found out how to repeatedly trigger this bug.

timminator commented:
I think I've come up with a way better solution now. As you mentioned already just making an exception for transcoded streams was not the right way to go.
Now I added a dictionary where the CloseIfNeededAsync task saves the sessionid for which it is called.
The OnPlaybackStopped task is then checking if the dictionary already has an entry with the same sessionid. If that is the case the CloseLivestream call is skipped.
This resolves the issue aswell. Could you take another look at my new commit?

gnattu commented:
This is still a bit leaky to me because it assumes all items has to be closed twice, and if it does not the livestream id is going to be stored forever. Can it be handled a bit more like `_activeConnections` which is added when open and removed when closing, but we just remove the id immediately after the handler is called?

cvium commented:
ConsumerCount shouldn't be an int. It should probably be a HashSet of session IDs. When the set is empty, close the stream.

timminator commented:
@gnattu: You were correct. My implementation from yesterday was indeed leaking.
I took another look at it.
PlaybackStopped will always be triggered and this task is now properly removing the entries out of the dictionary.
Unless Issue #13225 occurs, but I opened already another PR for that.
Furthermore i added a clear command in the "OnApplicationStopping" function.
For debugging i logged the entries inside the dictionary and they are now properly removed.

@cvium: That would also be an option. But that seems like a way bigger overhaul and would require quite a lot of changes. My new implentation seems to fix the issue aswell.

Could someone of you take another look at my updated commit?

timminator commented:
Hi! I came back to this issue yesterday and noticed an issue with my last version.
I read again through your messages @gnattu and noticed that last time I had not quite understood your suggestion:

> Can it be handled a bit more like _activeConnections which is added when open and removed when closing, but we just remove the id immediately after the handler is called?

I tried this and it is perfect! It works correctly and does not require changing the ConsumerCount type. I updated my commit where I implemented your approach now. The dictionary saves the sessionids for a livestream on Playbackstart and CloseIfNeededAsync() and OnPlaybackStopped() remove them immediately. The CloseLiveStream() function is only run when the sessionId was still in the dictionary.

I would really appreciate it if one of you or both could review this PR because this issue is affecting me quite often. :-)

crobibero commented on Emby.Server.Implementations/Session/SessionManager.cs:
```diff
@@ -62,6 +62,8 @@ public sealed class SessionManager : ISessionManager, IAsyncDisposable
         private readonly ConcurrentDictionary<string, SessionInfo> _activeConnections
             = new(StringComparer.OrdinalIgnoreCase);
 
+        private readonly ConcurrentDictionary<string, HashSet<string>> _activeLiveStreamSessions = new();
```
```suggestion
        private readonly ConcurrentDictionary<string, HashSet<string>> _activeLiveStreamSessions = new(StringComparer.OrdinalIgnoreCase);
```

crobibero commented on Emby.Server.Implementations/Session/SessionManager.cs:
```diff
@@ -309,7 +311,36 @@ public async Task CloseIfNeededAsync(SessionInfo session)
                 if (!string.IsNullOrEmpty(session.PlayState?.LiveStreamId))
                 {
-                    await _mediaSourceManager.CloseLiveStream(session.PlayState.LiveStreamId).ConfigureAwait(false);
+                    bool liveStreamNeedsToBeClosed = false;
```
Now that this is more then 1 line please extract it (and L1046-L1078) to a common function

timminator commented:
Thank you crobibero for approving. @gnattu @cvium Could you also take a look at my PR after my recent changes?

JPVenson commented on Emby.Server.Implementations/Session/SessionManager.cs:
```diff
@@ -62,6 +62,9 @@ public sealed class SessionManager : ISessionManager, IAsyncDisposable
         private readonly ConcurrentDictionary<string, SessionInfo> _activeConnections
             = new(StringComparer.OrdinalIgnoreCase);
 
+        private readonly ConcurrentDictionary<string, HashSet<string>> _activeLiveStreamSessions
```
Why not use a ConcurrentDictionary as value and use that for locking too instead of the Lock statement?

timminator replied:
Sorry for bothering. Do you have any thoughts on this?

JPVenson replied:
Yes that looks good.
Either you use a ConcurrentDictionary (as dotnet dont really have a ThreadSafe random access list) or you rebuild that behavior yourself using a ReaderWriterLock but that would bring other issues from an archtectual standpoint.

So a `ConcurrentDictionary<string, object>` is the easiest/simplest choice here

timminator replied:
Thank you for your answer. I've updated it again. Please take another look at it.

cvium replied:
> So a ConcurrentDictionary<string, object> is the easiest/simplest choice here

Normally you would use byte as the value type as a stand-in for a concurrent list

timminator commented:
Hi!
I've noticed a very closely related issue yesterday, when implementing the requested changes by JPVenson. I've opened another issue #13721 for that. It boils down to the fact that the same issue also occurs when a transcoding kill timer is reached.

The problem for me was, that the transcodingjob has no info about the sessionid, it has only a playsessionid, so I could not just let it use the newly added function "CloseLivestreamIfNeededAsync".  I also found no easy way to pass the sessionId to the transcoding job.

But I think I found a good solution. OnPlaybackStart and OnPlaybackProgress have info about the current sessionId and PlaysessionId at the same time, so they can be associated with each other and also be updated in case the playsessionid changes.
I've updated the ConcurrentDictionary to store both ids and updated CloseLivestreamIfNeededAsync to accept the sessionId or the playsessionid. This meant that I could use now this function also in the KilltranscodingJob Task. That required making the newly added function public. That explains the ABI differences.

With my newly added update it works now correctly and the log does not show the the second reduction of the consumercount (in comparison to the log in #13721).

```
[16:15:27] [INF] [14] Emby.Server.Implementations.Library.MediaSourceManager: Live stream a30d2ab40a1c24c15fb07e9537d38205 consumer count is now 1
[16:15:28] [INF] [14] Emby.Server.Implementations.HttpServer.WebSocketManager: WS 127.0.0.1 closed
[16:16:27] [INF] [16] MediaBrowser.MediaEncoding.Transcoding.TranscodeManager: Transcoding kill timer stopped for JobId 8965e52803824d62869eca71e6503cef PlaySessionId 47027a43da7a4af8a59ca5bd4141f0e8. Killing transcoding
[16:16:27] [INF] [16] MediaBrowser.Controller.MediaEncoding.TranscodingJob: Stopping ffmpeg process with q command for C:\Users\GPUVM\AppData\Local\jellyfin\cache\transcodes\665ec621e7fa0e6ebda80224b1d287a3.m3u8
[16:16:32] [INF] [16] MediaBrowser.Controller.MediaEncoding.TranscodingJob: Killing FFmpeg process for C:\Users\GPUVM\AppData\Local\jellyfin\cache\transcodes\665ec621e7fa0e6ebda80224b1d287a3.m3u8
[16:16:32] [INF] [16] MediaBrowser.MediaEncoding.Transcoding.TranscodeManager: Deleting partial stream file(s) C:\Users\GPUVM\AppData\Local\jellyfin\cache\transcodes\665ec621e7fa0e6ebda80224b1d287a3.m3u8
[16:16:32] [ERR] [14] MediaBrowser.MediaEncoding.Transcoding.TranscodeManager: FFmpeg exited with code -1
```

If you have any issues with this update of mine for fixing the second issue I can also backroll this commit to the previous state and we can fix Issue #13721 with a new PR.
I will request a new review from you @crobibero because there are now more changes since your last review. I hope that is okay. :-)

timminator commented:
Thank you JPVenson and crobibero for approving my PR! Would it be possible to merge it soon?
